### PR TITLE
Base our Recycler Adapters on Android's ListAdapter

### LINKED ISSUE #5925: Move all our adapters to ListAdapter base class
## 🚧 Refactor Proposal

Related slack thread https://toggl.slack.com/archives/C02APMYTM/p1568810483151900

Our implementation of `BaseRecyclerAdapter` is complicated and there are now several issues that might be related to some unpredictable behaviors (#5692 #5691 ). Therefore, we want to move to standard `ListAdapter` [docs](https://developer.android.com/reference/android/support/v7/recyclerview/extensions/ListAdapter) and let google handle all of these issues.

## Dependencies

First, we need to verify that `ListAdapter` is available and works correctly on Xamarin.

## Technical aspects

Basically we need to replace `BaseRecyclerAdapter` everywhere as a base class and use the `ListAdapter` instead.  Extending the `ListAdapter` is pretty straight forward, just follow the [docs](https://developer.android.com/reference/android/support/v7/recyclerview/extensions/ListAdapter)

### PR DESCRIPTION
**Pull request by semanticer**
## What's this?

We move from our custom implementation of `BaseRecyclerAdapter` to implementation that delegates much of is logic to Android's `ListAdapter`

### Relationships

Closes #5925 

Maybe also #5692 and #5820 - these are hard to reproduce and verify but there is a good chance they are fixed since the implementation completely changed towards standard Android way

## Why do we want this?

We want to use existing battle-tested solution instead of our DIY implementation

## How is it done?

`BaseRecyclerAdapter` now extends `ListAdapter` and bridges it's methods.

### Why not another way?

Let me know if you have any other ideas

### Side effects

Since `ListAdapter` returns java objects we need to cast them to our C# equivalents.  `BaseRecyclerAdapter` now provides 'safe' methods like `getTypedItem(pos)` in order to shield all its subclasses from this casting magic shenanigans. This can theoretically be a performance problem, especially in the DiffingCallback.

## Review hints

Well if you have time you can double-check all our recycler views are working and updating the same way as before. 

## :squid: Permissions

Me or @heytherewill

### CODE REVIEW

File: Toggl.Droid/Extensions/ObjectTypeExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+namespace Toggl.Droid.Extensions
+{
+    public static class ObjectTypeExtensions
+    {
+        public static T Cast<T>(this Java.Lang.Object obj) where T : class
+        {
+            var propertyInfo = obj.GetType().GetProperty("Instance");
```
Comment (heytherewill) [inline Toggl.Droid/Extensions/ObjectTypeExtensions.cs]: Shit, is this needed? Won't `obj as T` suffice?

Alternatively, couldn't you simply `obj.JavaCast<T>`? I don't really remember what JavaCast does, I'm just trying to figure out a solution that doesn't involve using reflection 🙈

File: Toggl.sln [hunk 2]
```diff
@@ -21,29 +21,29 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Toggl.Shared.Tests", "Toggl
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Toggl.Networking", "Toggl.Networking\Toggl.Networking.csproj", "{5ACEB155-9DD4-4309-BCC9-997A09950393}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Toggl.Networking.Tests", "Toggl.Networking.Tests\Toggl.Networking.Tests.csproj", "{8598D5CB-4E3A-4F99-8E80-CC4EA8609ED8}"
```
Comment (heytherewill) [inline Toggl.sln]: Not relevant to this PR

File: Toggl.Droid/Extensions/ObjectTypeExtensions.cs (hunk 1 above)
Comment (kornelijepetak) [reply Toggl.Droid/Extensions/ObjectTypeExtensions.cs]: I am almost 100% sure that this can be done with `as T`.
Avoid reflection as hell 😛

Comment (kornelijepetak) [reply Toggl.Droid/Extensions/ObjectTypeExtensions.cs]: Especially since this code here will be called quite a lot!

Comment (semanticer) [reply Toggl.Droid/Extensions/ObjectTypeExtensions.cs]: `obj as T` results in `null` :-( 

I will try the `JavaCast` variant

Comment (semanticer) [reply Toggl.Droid/Extensions/ObjectTypeExtensions.cs]: `obj.JavaCast<T>` requires T to implement `IJavaObject` https://docs.microsoft.com/en-gb/dotnet/api/Android.Runtime.Extensions.JavaCast?view=xamarin-android-sdk-9

**Comment by pdrj** [comment]
Regarding #5820 - good news - I can't reproduce it with data that crashes each time on develop :).

Will test the other adapters too.

**Comment by semanticer** [comment]
After discussion here https://toggl.slack.com/archives/CD40QCV25/p1569248243020200 and the bug fund by @pdrj above ☝️ we decided to move the related issue to the next release and try some other approach with 0 reflection involved.

**Comment by heytherewill** [comment]
Is it worth to keep this open?

**Comment by semanticer** [comment]
We agreed that reflection is unacceptable and this solution depends on it. I'm closing this PR, should try a different approach. 
Related slack thread: https://toggl.slack.com/archives/CD40QCV25/p1569248243020200

**Pull request closed without merging**